File: DA/python/load_campaign_spend_to_warehouse.py
"""
Load campaign spend raw CSVs into the warehouse.
Handles schema drift: Q1 has (Campaign ID, Date, Spend, Impressions, Clicks),
Q2 has (campaign_id, date, spend_usd, impr, click_count).
Skips footer lines; normalizes dates; empty Spend -> 0.

Run from project root: py python/load_campaign_spend_to_warehouse.py
"""

import csv
import sqlite3
import re
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent
RAW_Q1 = PROJECT_ROOT / "raw" / "vendor" / "campaign_spend_2024q1.csv"
RAW_Q2 = PROJECT_ROOT / "raw" / "vendor" / "campaign_spend_2024q2.csv"
WAREHOUSE_DB = PROJECT_ROOT / "warehouse" / "brightwave.sqlite"
SCHEMA_SQL = PROJECT_ROOT / "warehouse" / "schema.sql"
SEED_SQL = PROJECT_ROOT / "warehouse" / "seed_dimensions.sql"
SEED_CAMPAIGNS_SQL = PROJECT_ROOT / "warehouse" / "seed_campaigns.sql"


def normalize_date(value):
    """Parse common date formats to YYYY-MM-DD. Returns None if invalid."""
    if value is None or (isinstance(value, str) and value.strip() == ""):
        return None
    s = str(value).strip()
    m = re.match(r"^(\d{4})-(\d{2})-(\d{2})", s)
    if m:
        return f"{m.group(1)}-{m.group(2)}-{m.group(3)}"
    m = re.match(r"^(\d{1,2})/(\d{1,2})/(\d{4})$", s)
    if m:
        mo, day, yr = int(m.group(1)), int(m.group(2)), int(m.group(3))
        return f"{yr}-{mo:02d}-{day:02d}"
    return None


def safe_float(value):
    if value is None:
        return None
    if isinstance(value, (int, float)):
        return float(value)
    s = str(value).strip()
    if s == "" or s.lower() == "null":
        return None
    try:
        return float(s)
    except (ValueError, TypeError):
        return None


def safe_int(value):
    if value is None:
        return None
    if isinstance(value, int):
        return value
    s = str(value).strip()
    if s == "" or s.lower() == "null":
        return None
    try:
        return int(float(s))
    except (ValueError, TypeError):
        return None


def run_sql_file(conn, path):
    with open(path, "r", encoding="utf-8") as f:
        conn.executescript(f.read())


def read_q1(path):
    """Q1: Campaign ID, Date, Spend, Impressions, Clicks. Skip footer."""
    rows = []
    with open(path, "r", encoding="utf-8", newline="") as f:
        reader = csv.DictReader(f)
        for r in reader:
            cid = (r.get("Campaign ID") or "").strip()
            if not cid.startswith("CAMP-"):
                continue
            date_iso = normalize_date(r.get("Date"))
            if not date_iso:
                continue
            spend = safe_float(r.get("Spend"))
            impressions = safe_int(r.get("Impressions"))
            clicks = safe_int(r.get("Clicks"))
            rows.append({
                "campaign_id": cid,
                "date": date_iso,
                "spend": spend if spend is not None else 0.0,
                "impressions": impressions if impressions is not None else 0,
                "clicks": clicks if clicks is not None else 0,
            })
    return rows


def read_q2(path):
    """Q2: campaign_id, date, spend_usd, impr, click_count."""
    rows = []
    with open(path, "r", encoding="utf-8", newline="") as f:
        reader = csv.DictReader(f)
        for r in reader:
            cid = (r.get("campaign_id") or "").strip()
            if not cid.startswith("CAMP-"):
                continue
            date_iso = normalize_date(r.get("date"))
            if not date_iso:
                continue
            spend = safe_float(r.get("spend_usd"))
            impressions = safe_int(r.get("impr"))
            clicks = safe_int(r.get("click_count"))
            rows.append({
                "campaign_id": cid,
                "date": date_iso,
                "spend": spend if spend is not None else 0.0,
                "impressions": impressions if impressions is not None else 0,
                "clicks": clicks if clicks is not None else 0,
            })
    return rows


def main():
    WAREHOUSE_DB.parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(WAREHOUSE_DB)

    run_sql_file(conn, SCHEMA_SQL)
    run_sql_file(conn, SEED_SQL)
    run_sql_file(conn, SEED_CAMPAIGNS_SQL)
    conn.commit()

    all_rows = []
    if RAW_Q1.exists():
        all_rows.extend(read_q1(RAW_Q1))
    if RAW_Q2.exists():
        all_rows.extend(read_q2(RAW_Q2))

    # Dedupe by (campaign_id, date) — keep last (Q2 overwrites Q1 if same key)
    seen = {}
    for row in all_rows:
        key = (row["campaign_id"], row["date"])
        seen[key] = row
    unique = list(seen.values())

    cur = conn.cursor()
    for row in unique:
        cur.execute(
            """
            INSERT OR REPLACE INTO fact_campaign_spend_daily
            (campaign_id, date, spend, impressions, clicks)
            VALUES (?, ?, ?, ?, ?)
            """,
            (row["campaign_id"], row["date"], row["spend"], row["impressions"], row["clicks"]),
        )
    conn.commit()
    conn.close()
    print(f"Loaded {len(unique)} campaign spend rows into fact_campaign_spend_daily.")


if __name__ == "__main__":
    main()

File: DA/r/load_campaign_spend_to_warehouse.R
# Load campaign spend raw CSVs into the warehouse.
# Handles schema drift: Q1 (Campaign ID, Date, Spend, ...) vs Q2 (campaign_id, date, spend_usd, impr, click_count).
# Required: DBI, RSQLite (jsonlite not needed for this script)
# Run from project root: Rscript r/load_campaign_spend_to_warehouse.R

library(DBI)
library(RSQLite)

# Paths (assume working directory is project root)
project_root <- getwd()
if (basename(project_root) == "r") project_root <- dirname(project_root)
raw_q1 <- file.path(project_root, "raw", "vendor", "campaign_spend_2024q1.csv")
raw_q2 <- file.path(project_root, "raw", "vendor", "campaign_spend_2024q2.csv")
warehouse_db <- file.path(project_root, "warehouse", "brightwave.sqlite")
schema_sql <- file.path(project_root, "warehouse", "schema.sql")
seed_sql <- file.path(project_root, "warehouse", "seed_dimensions.sql")
seed_campaigns_sql <- file.path(project_root, "warehouse", "seed_campaigns.sql")

run_sql_file <- function(conn, path) {
  sql <- paste(readLines(path, encoding = "UTF-8"), collapse = "\n")
  statements <- strsplit(sql, ";\\s*\n")[[1]]
  for (s in statements) {
    s <- trimws(s)
    if (nchar(s) > 0) dbExecute(conn, s)
  }
}

normalize_date <- function(value) {
  if (is.null(value) || (is.character(value) && trimws(value) == ""))
    return(NA_character_)
  s <- trimws(as.character(value))
  if (grepl("^[0-9]{4}-[0-9]{2}-[0-9]{2}", s)) return(substr(s, 1, 10))
  d <- as.Date(s, format = "%m/%d/%Y")
  if (!is.na(d)) return(format(d, "%Y-%m-%d"))
  NA_character_
}

safe_float <- function(x) {
  if (is.null(x)) return(NA_real_)
  if (is.character(x) && trimws(x) %in% c("", "null", "NULL")) return(NA_real_)
  as.numeric(x)
}

safe_int <- function(x) {
  if (is.null(x)) return(NA_integer_)
  if (is.character(x) && trimws(x) %in% c("", "null", "NULL")) return(NA_integer_)
  as.integer(round(as.numeric(x)))
}

# Read Q1 CSV: Campaign ID, Date, Spend, Impressions, Clicks (skip footer)
read_q1 <- function(path) {
  if (!file.exists(path)) return(data.frame())
  d <- read.csv(path, stringsAsFactors = FALSE)
  d$campaign_id <- trimws(d$Campaign.ID)
  d <- d[grepl("^CAMP-", d$campaign_id), , drop = FALSE]
  d$date <- vapply(d$Date, function(x) normalize_date(x), character(1))
  d <- d[!is.na(d$date), , drop = FALSE]
  spend <- vapply(d$Spend, safe_float, numeric(1))
  impressions <- vapply(d$Impressions, safe_int, integer(1))
  clicks <- vapply(d$Clicks, safe_int, integer(1))
  data.frame(
    campaign_id = d$campaign_id,
    date = d$date,
    spend = ifelse(is.na(spend), 0, spend),
    impressions = ifelse(is.na(impressions), 0L, impressions),
    clicks = ifelse(is.na(clicks), 0L, clicks),
    stringsAsFactors = FALSE
  )
}

# Read Q2 CSV: campaign_id, date, spend_usd, impr, click_count
read_q2 <- function(path) {
  if (!file.exists(path)) return(data.frame())
  d <- read.csv(path, stringsAsFactors = FALSE)
  d$campaign_id <- trimws(d$campaign_id)
  d <- d[grepl("^CAMP-", d$campaign_id), , drop = FALSE]
  d$date_norm <- vapply(d$date, function(x) normalize_date(x), character(1))
  d <- d[!is.na(d$date_norm), , drop = FALSE]
  spend <- vapply(d$spend_usd, safe_float, numeric(1))
  impressions <- vapply(d$impr, safe_int, integer(1))
  clicks <- vapply(d$click_count, safe_int, integer(1))
  data.frame(
    campaign_id = d$campaign_id,
    date = d$date_norm,
    spend = ifelse(is.na(spend), 0, spend),
    impressions = ifelse(is.na(impressions), 0L, impressions),
    clicks = ifelse(is.na(clicks), 0L, clicks),
    stringsAsFactors = FALSE
  )
}

# ---- Main ----

dir.create(dirname(warehouse_db), recursive = TRUE, showWarnings = FALSE)
conn <- dbConnect(RSQLite::SQLite(), warehouse_db)

run_sql_file(conn, schema_sql)
run_sql_file(conn, seed_sql)
run_sql_file(conn, seed_campaigns_sql)

df1 <- read_q1(raw_q1)
df2 <- read_q2(raw_q2)
all_rows <- rbind(df1, df2)

# Dedupe by (campaign_id, date) — keep last
all_rows <- all_rows[!duplicated(all_rows[, c("campaign_id", "date")], fromLast = TRUE), ]

for (i in seq_len(nrow(all_rows))) {
  r <- all_rows[i, ]
  dbExecute(conn,
    "INSERT OR REPLACE INTO fact_campaign_spend_daily (campaign_id, date, spend, impressions, clicks) VALUES (?, ?, ?, ?, ?)",
    params = list(r$campaign_id, r$date, r$spend, r$impressions, r$clicks))
}

dbDisconnect(conn)
message("Loaded ", nrow(all_rows), " campaign spend rows into fact_campaign_spend_daily.")
